### PR TITLE
fix: doubled move_x accounting

### LINKED ISSUE #183: move_x is applied twice in check_screen_switch function
This is just a minor thing (and I'm currently on a train so I can't verify this), but I noticed one thing:

```cpp
void update_mouse_position(device_t *state, mouse_values_t *values) {
	// snip…

  /* Calculate movement */
  int offset_x = accelerate(values->move_x) * (current->speed_x >> reduce_speed);
  int offset_y = accelerate(values->move_y) * (current->speed_y >> reduce_speed);

  /* Update movement */
  state->pointer_x = move_and_keep_on_screen(state->pointer_x, offset_x);
  state->pointer_y = move_and_keep_on_screen(state->pointer_y, offset_y);

	// snip…
}
```

In `update_mouse_position` we already set the new value for `state->pointer_x`.

```cpp
void check_screen_switch(const mouse_values_t *values, device_t *state) {
  int new_x        = state->pointer_x + values->move_x;
	// snip…
}
```

And then in `check_screen_switch` we add `values->move_x` to `state->pointer_x` again. Isn't `values->move_x` added twice now?

_Originally posted by @jalmeroth in https://github.com/hrvach/deskhop/issues/3#issuecomment-2520608664_

### PR DESCRIPTION
**Pull request by jalmeroth**
Fixes #183

### CODE REVIEW

**Comment by jalmeroth** [comment]
Also this **adds** the `jump_treshold` to `MIN_SCREEN_COORD` and **substracts** it from `MAX_SCREEN_COORD`

File: src/mouse.c [hunk 1]
```diff
@@ -183,11 +183,10 @@ void switch_virtual_desktop(device_t *state, output_t *output, int new_index, in
           )___(          )___(     |     )___(          )___(          )___(
 */
 void check_screen_switch(const mouse_values_t *values, device_t *state) {
-    int new_x        = state->pointer_x + values->move_x;
     output_t *output = &state->config.output[state->active_output];
 
-    bool jump_left  = new_x < MIN_SCREEN_COORD - state->config.jump_treshold;
-    bool jump_right = new_x > MAX_SCREEN_COORD + state->config.jump_treshold;
+    bool jump_left  = state->pointer_x <= MIN_SCREEN_COORD + state->config.jump_treshold;
```
Comment (hrvach) [inline src/mouse.c:188]: If you move the mouse to MIN_SCREEN_COORD on one device, you'd get transferred to MAX_SCREEN_COORD on the other and immediately get transferred back?

Also, this effectively leaves you without a way to ever click on the first and last pixel columns on their respective screens?

Comment (jalmeroth) [reply src/mouse.c:188]: Hmn, okay… but can you even go below MIN_SCREEN_COORD (0) or further then MAX_SCREEN_COORD?

**Comment by hrvach** [comment]
> Also this **adds** the `jump_treshold` to `MIN_SCREEN_COORD` and **substracts** it from `MAX_SCREEN_COORD`, so we actually jump between `jump_treshold` and `min|max`

The idea behind jump_threshold is - you can make the jump less accidental by making sort of a "barrier" between screens. If you gently move the mouse, nothing happens but a sudden movement (which results in a high delta x) would be sufficient to exceed the screen edge coordinate + some offset that moves the MIN coord further to the left or MAX coord further to the right. 

You're right about the code first doing the "add the X offset" and then checking "would adding x again to the current position get us across to the other screen", that is a crappy solution. It would maybe make most sense if move_and_keep_on_screen() would signal if there was an overflow and use that information later?

File: src/mouse.c (hunk 1 above)
Comment (hrvach) [reply src/mouse.c:188]: Physically - no, but you could request it and that in theory should trigger screen switch if there is a screen to switch to, or put you at the very edge if there isn't.

New_x is a bad variable name, in this case it's desired_x_coordinate, it doesn't get assigned to anything but is merely used to determine if you tried to leave a screen.

**Comment by jalmeroth** [comment]
> The idea behind jump_threshold is - you can make the jump less accidental by making sort of a "barrier" between screens. If you gently move the mouse, nothing happens but a sudden movement (which results in a high delta x) would be sufficient to exceed the screen edge coordinate + some offset that moves the MIN coord further to the left or MAX coord further to the right.

Ok, now I understand its purpose better, but the mouse has already been moved (report has been sent to PC a) at this point, right? So we are actively jumping to PC B afterwards, but then we wait for the next report to come in.

**Comment by hrvach** [comment]
> Ok, now I understand its purpose better, but the mouse has already been moved (report has been sent to PC A) at this point, right? So we are actively jumping to PC B afterwards, but then we wait for the next report to come in.

Yes, the logic is not 100% correct there. The move part seems OK as it moves either how much you wanted (or possible) towards the edge, send the report and then check to see if we need to switch ... but that part is a bit flawed, it should probably use the information from update_mouse_position() call to see if that overflowed or not and avoid using the new X location + raw offset.

**Comment by jalmeroth** [comment]
So, how do we proceed? Initially I thought this would be a quick-win, but it seems to be more complex then this. Shall I close the PR or would like to build on it?

**Comment by hrvach** [comment]
> So, how do we proceed? Initially I thought this would be a quick-win, but it seems to be more complex then this. Shall I close the PR or would like to build on it?

With my luck, there are no quick wins :) Since you discovered the issue (well done), we should use that momentum to fix it. I'll try building on top of your PR, thanks for making it!

**Comment by hrvach** [comment]
I think this should be fixed now with the latest commit.

**Comment by jalmeroth** [comment]
You rock @hrvach! 🎸 Thank you very much for going the extra mile.

**Pull request closed without merging**